File: src/a_star.zig
const std = @import("std");
const zm = @import("zmath");
const math = std.math;
const Order = math.Order;
const FLOAT_MAX = std.math.f32_max;
const Allocator = std.mem.Allocator;
const DungeonTile = @import("dungeon.zig").DungeonTile;
const ArrayList = std.ArrayList;

const Node = struct {
    position: zm.Vec,
    parentPosition: zm.Vec,
    gCost: f32,
    hCost: f32,
    fCost: f32,
};

pub fn aStar(allocator: *const Allocator, map: []const []const DungeonTile, start: zm.Vec, goal: zm.Vec) !ArrayList(zm.Vec) {
    if (vecEquals(start, goal)) {
        return ArrayList(zm.Vec).init(allocator.*);
    }

    var nodeMap = try initNodeMap(allocator, map.len);
    defer {
        for (nodeMap) |row| {
            allocator.free(row);
        }
        allocator.free(nodeMap);
    }
    var closedList = try initClosedList(allocator, map.len);
    defer {
        for (closedList) |row| {
            allocator.free(row);
        }
        allocator.free(closedList);
    }

    var startNode = nodeMap[@floatToInt(usize, start[1])][@floatToInt(usize, start[0])];
    startNode.position = start;
    startNode.fCost = 0.0;
    startNode.gCost = 0.0;
    startNode.hCost = 0.0;

    var openList = ArrayList(Node).init(allocator.*);
    defer openList.deinit();

    try openList.append(startNode);

    while (openList.items.len != 0) {
        // Pop element with min cost
        var idx: usize = 0;
        var minIdx: usize = 0;
        var minCost: f32 = 0;
        while (idx < openList.items.len) : (idx += 1) {
            var tmpNode = openList.items[idx];
            if (tmpNode.fCost < minCost) {
                minCost = tmpNode.fCost;
                minIdx = idx;
            }
        }
        var currentNode = openList.orderedRemove(minIdx);
        var currentX = @floatToInt(usize, currentNode.position[0]);
        var currentY = @floatToInt(usize, currentNode.position[1]);

        closedList[currentY][currentX] = true;

        var neighbors = getNeighbors(nodeMap, currentNode);
        for (neighbors) |neighbor| {
            var neighborX = @floatToInt(usize, neighbor.position[0]);
            var neighborY = @floatToInt(usize, neighbor.position[1]);
            if (isWalkable(map, neighbor)) {
                if (vecEquals(neighbor.position, goal)) {
                    var nodeToUpdate = &(nodeMap[neighborY][neighborX]);
                    nodeToUpdate.parentPosition = currentNode.position;

                    return makePath(allocator, nodeMap, goal);
                } else if (closedList[neighborY][neighborX] == false) {
                    var gNew = currentNode.gCost + 1.0;
                    var hNew = heuristic(neighbor.position, goal);
                    var fNew = gNew + hNew;
                    var nodeToUpdate = &(nodeMap[neighborY][neighborX]);
                    if (nodeToUpdate.fCost == FLOAT_MAX or nodeMap[neighborY][neighborX].fCost > fNew) {
                        nodeToUpdate.fCost = fNew;
                        nodeToUpdate.gCost = gNew;
                        nodeToUpdate.hCost = hNew;
                        nodeToUpdate.parentPosition = currentNode.position;
                        try openList.append(nodeToUpdate.*);
                    }
                }
            }
        }
    }

    return ArrayList(zm.Vec).init(allocator.*);
}

fn initClosedList(allocator: *const Allocator, size: usize) ![][]bool {
    var x: usize = 0;
    var y: usize = 0;
    var closedList = ArrayList([]bool).init(allocator.*);
    defer closedList.deinit();

    while (y < size) : (y += 1) {
        var row = ArrayList(bool).init(allocator.*);
        defer row.deinit();
        x = 0;
        while (x < size) : (x += 1) {
            try row.append(false);
        }
        try closedList.append(row.toOwnedSlice());
    }

    return closedList.toOwnedSlice();
}

fn initNodeMap(allocator: *const Allocator, size: usize) ![][]Node {
    var x: usize = 0;
    var y: usize = 0;
    var nodeMap = ArrayList([]Node).init(allocator.*);
    defer nodeMap.deinit();

    while (y < size) : (y += 1) {
        var row = ArrayList(Node).init(allocator.*);
        defer row.deinit();
        x = 0;
        while (x < size) : (x += 1) {
            var node = Node{
                .fCost = FLOAT_MAX,
                .gCost = FLOAT_MAX,
                .hCost = FLOAT_MAX,
                .parentPosition = zm.f32x4(-1.0, -1.0, 0.0, 0.0),
                .position = zm.f32x4(@intToFloat(f32, x), @intToFloat(f32, y), 0.0, 0.0),
            };
            try row.append(node);
        }
        try nodeMap.append(row.toOwnedSlice());
    }

    return nodeMap.toOwnedSlice();
}

fn getNeighbors(map: [][]Node, node: Node) [4]Node {
    var vec = node.position;
    var x = @floatToInt(usize, vec[0]);
    var y = @floatToInt(usize, vec[1]);

    return [_]Node{
        map[y + 1][x],
        map[y - 1][x],
        map[y][x + 1],
        map[y][x - 1],
    };
}

fn isWalkable(map: []const []const DungeonTile, node: Node) bool {
    var vec = node.position;
    var tile = map[@floatToInt(usize, vec[1])][@floatToInt(usize, vec[0])];
    return tile != .empty and tile != .wall;
}

fn vecEquals(vec1: zm.Vec, vec2: zm.Vec) bool {
    return vec1[0] == vec2[0] and
        vec1[1] == vec2[1];
}

fn heuristic(v1: zm.Vec, v2: zm.Vec) f32 {
    return math.fabs(v1[0] - v2[0]) + math.fabs(v1[1] - v2[1]);
}

fn makePath(allocator: *const Allocator, map: [][]Node, goal: zm.Vec) !ArrayList(zm.Vec) {
    var path = ArrayList(zm.Vec).init(allocator.*);
    defer path.deinit();

    var x = @floatToInt(usize, goal[0]);
    var y = @floatToInt(usize, goal[1]);

    var node = map[y][x];
    while (node.parentPosition[0] != -1.0 and node.parentPosition[1] != -1.0) {
        node = map[y][x];
        var parent = node.parentPosition;

        if (parent[0] == -1.0 and parent[1] == -1.0) {
            break;
        }
        try path.append(node.position);

        var parentX = @floatToInt(usize, parent[0]);
        var parentY = @floatToInt(usize, parent[1]);
        x = parentX;
        y = parentY;
    }

    var reversePath = ArrayList(zm.Vec).init(allocator.*);

    var i: i32 = @intCast(i32, path.items.len) - 1;
    while (i > 0) : (i -= 1) {
        try reversePath.append(path.items[@intCast(usize, i)]);
    }

    return reversePath;
}

File: src/a_star.cpp
#include "hero.h"
#include "dungeon.h"
#include <glm/vec2.hpp>
#include <vector>
#include "a_star.h"
#include <array>
#include <cfloat>
#include "dungeon_tile.h"
#include <iostream>
#include <stack>
#include <algorithm>

struct Node {
	int y;
	int x;
	int parentX;
	int parentY;
	float gCost;
	float hCost;
	float fCost;
};

bool node_cmp(const Node &a, const Node &b) {
	return a.fCost < b.fCost;
}

float mhtndist(glm::vec2 v1, glm::vec2 v2) {
	return std::abs(v1.x - v2.x) + std::abs(v1.y - v2.y);
}

bool is_walkable(Dungeon &dungeon, glm::vec2 v) {
	int x = (int)v.x;
	int y = (int)v.y;
	DungeonTile tile = dungeon.map[y][x];

	return (tile == DungeonTile::Floor ||
			tile == DungeonTile::Door ||
			tile == DungeonTile::Corridor) &&
		dungeon.enemies[y][x] == 0;
}

bool are_same_position(glm::vec2 v1, glm::vec2 v2) {
	return v1.x == v2.x && v1.y == v2.y;
}

void init_data(Dungeon &dungeon,
			   std::vector<std::vector<Node>> &nodes,
			   std::vector<std::vector<bool>> &visited_nodes) {
	nodes = std::vector<std::vector<Node>>(dungeon.size, std::vector<Node>(dungeon.size));
	visited_nodes = std::vector<std::vector<bool>>(dungeon.size, std::vector<bool>(dungeon.size));

	for (int y = 0; y < dungeon.size; y++) {
		for (int x = 0; x < dungeon.size; x++) {
			nodes[y][x].fCost = FLT_MAX;
			nodes[y][x].gCost = FLT_MAX;
			nodes[y][x].hCost = FLT_MAX;
			nodes[y][x].parentX = -1;
			nodes[y][x].parentY = -1;
			nodes[y][x].x = x;
			nodes[y][x].y = y;

			visited_nodes[y][x] = false;
		}
	}
}

std::vector<glm::vec2> makePath(std::vector<std::vector<Node>> map, glm::vec2 dest) {
	int x = dest.x;
	int y = dest.y;
	std::stack<glm::vec2> path;
	std::vector<glm::vec2> usablePath;

	while (x != -1 && y != -1)
	{
		Node node = map[y][x];
		path.push(glm::vec2(node.x, node.y));
		x = node.parentX;
		y = node.parentY;
	}

	while (!path.empty()) {
		glm::vec2 top = path.top();
		path.pop();
		usablePath.emplace_back(top);
	}

	return usablePath;
}

std::vector<glm::vec2> AStar::find_path(Dungeon &dungeon, glm::vec2 &start, glm::vec2 &goal) {
	std::vector<glm::vec2> empty;

#ifdef DEBUG
	printf("Starting A* from (%.1f, %.1f) to (%.1f, %.1f)\n", start[0], start[1], goal[0], goal[1]);
#endif

	if (are_same_position(start, goal)) {
		printf("Start and goal are the same\n");
		return empty;
	}

	std::vector<std::vector<bool>> visited_nodes;
	std::vector<std::vector<Node>> nodes;

	init_data(dungeon, nodes, visited_nodes);

	int x = start.x;
	int y = start.y;
	nodes[y][x].fCost = 0.0;
	nodes[y][x].gCost = 0.0;
	nodes[y][x].hCost = 0.0;
	nodes[y][x].parentX = -1;
	nodes[y][x].parentY = -1;


	std::vector<Node> queue; // TODO: Use mean heap to search for data
	queue.emplace_back(nodes[y][x]);
	bool destinationFound = false;

	glm::vec2 neighbor_offset[] = {
		glm::vec2(1.0f, 0.0f),
		glm::vec2(0.0f, -1.0f),
		glm::vec2(-1.0f, 0.0f),
		glm::vec2(0.0f, 1.0f),
	};

	while (!queue.empty()) {
		auto node_it = std::min_element(queue.begin(), queue.end(), node_cmp);
		Node node = *node_it;

		queue.erase(node_it);
		x = node.x;
		y = node.y;

		visited_nodes[y][x] = true;
		for(auto offset : neighbor_offset) {
			int new_x = x + offset[0];
			int new_y = y + offset[1];
			Node &neighbor = nodes[new_y][new_x];
			double gNew = 0;
			double hNew = 0;
			double fNew = 0;

			if (is_walkable(dungeon, glm::vec2(new_x, new_y))) {

				if (are_same_position(glm::vec2(new_x, new_y), goal)) {
					printf("Found destination\n");
					neighbor.parentX = x;
					neighbor.parentY = y;
					destinationFound = true;
					return makePath(nodes, goal);
				}
				else if (!(visited_nodes[new_y][new_x])) {
					gNew = node.gCost + 1.0;
					hNew = mhtndist(glm::vec2(new_x, new_y), goal);
					fNew = gNew + hNew;

					if (fNew < neighbor.fCost) {
						neighbor.fCost = fNew;
						neighbor.gCost = gNew;
						neighbor.hCost = hNew;
						neighbor.parentX = x;
						neighbor.parentY = y;
						queue.emplace_back(neighbor);
					}
				}
			}
		}
	}
	if (destinationFound == false) {
		printf("Destination not found\n");
		return empty;
	}
	return empty;
}
